Can't build on MSYS2/MinGW-w64 (Windows)

Hi, @jgaa.

On my Windows computer, my development environment is based on [MSYS2 and MinGW-w64](http://www.msys2.org/).

CMake (`mingw-w64-x86_64-cmake`) and Boost (`mingw-w64-x86_64-boost`) are installed. I even installed rapidjson (`mingw-w64-x86_64-rapidjson`), just in case.

I also have tried to set CMAKE_CXX_FLAGS to `-std=c++14`, but nothing changed.

CMake's log:
```
-- The C compiler identification is GNU 6.3.0
-- The CXX compiler identification is GNU 6.3.0
-- Check for working C compiler: C:/msys64/mingw64/bin/cc.exe
-- Check for working C compiler: C:/msys64/mingw64/bin/cc.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: C:/msys64/mingw64/bin/c++.exe
-- Check for working CXX compiler: C:/msys64/mingw64/bin/c++.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Using C:/msys64/mingw64/bin/c++.exe
-- Found ZLIB: C:/msys64/mingw64/lib/libz.dll.a (found version "1.2.11")
-- Found OpenSSL: C:/msys64/mingw64/lib/libssl.dll.a (found version "1.0.2l")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE
-- Boost version: 1.64.0
-- Found the following Boost libraries:
--   system
--   program_options
--   filesystem
--   date_time
--   context
--   coroutine
--   chrono
--   log
--   thread
--   log_setup
--   regex
--   atomic
-- Writing the current configuration to C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/config.h
-- Configuring done
-- Generating done
-- Build files have been written to: C:/msys64/home/gjacquemin/restc-cpp/dbuild
```
`make`'s log:
```
Scanning dependencies of target restc-cpp
[  1%] Building CXX object src/CMakeFiles/restc-cpp.dir/ChunkedReaderImpl.cpp.obj
In file included from C:/msys64/mingw64/x86_64-w64-mingw32/include/minwindef.h:163:0,
                 from C:/msys64/mingw64/x86_64-w64-mingw32/include/windef.h:8,
                 from C:/msys64/mingw64/x86_64-w64-mingw32/include/windows.h:69,
                 from C:/msys64/mingw64/x86_64-w64-mingw32/include/winsock2.h:23,
                 from C:/msys64/mingw64/include/boost/asio/detail/socket_types.hpp:33,
                 from C:/msys64/mingw64/include/boost/asio/detail/win_tss_ptr.hpp:23,
                 from C:/msys64/mingw64/include/boost/asio/detail/tss_ptr.hpp:25,
                 from C:/msys64/mingw64/include/boost/asio/detail/call_stack.hpp:20,
                 from C:/msys64/mingw64/include/boost/asio/impl/handler_alloc_hook.ipp:19,
                 from C:/msys64/mingw64/include/boost/asio/handler_alloc_hook.hpp:80,
                 from C:/msys64/mingw64/include/boost/asio/detail/handler_alloc_helpers.hpp:21,
                 from C:/msys64/mingw64/include/boost/asio/detail/bind_handler.hpp:19,
                 from C:/msys64/mingw64/include/boost/asio/detail/wrapped_handler.hpp:18,
                 from C:/msys64/mingw64/include/boost/asio/io_service.hpp:24,
                 from C:/msys64/mingw64/include/boost/asio/basic_io_object.hpp:19,
                 from C:/msys64/mingw64/include/boost/asio/basic_socket.hpp:20,
                 from C:/msys64/mingw64/include/boost/asio/basic_datagram_socket.hpp:20,
                 from C:/msys64/mingw64/include/boost/asio.hpp:21,
                 from C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:16,
                 from C:/msys64/home/gjacquemin/restc-cpp/src/ChunkedReaderImpl.cpp:3:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:116:9: error: expected identifier before '(' token
         DELETE
         ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:116:9: error: expected '}' before '(' token
In file included from C:/msys64/mingw64/x86_64-w64-mingw32/include/_mingw.h:12:0,
                 from C:/msys64/mingw64/x86_64-w64-mingw32/include/crtdefs.h:10,
                 from C:/msys64/mingw64/x86_64-w64-mingw32/include/assert.h:15,
                 from C:/msys64/home/gjacquemin/restc-cpp/src/ChunkedReaderImpl.cpp:1:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:116:9: error: expected unqualified-id before numeric constant
         DELETE
         ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:116:9: error: expected ')' before numeric constant
In file included from C:/msys64/home/gjacquemin/restc-cpp/src/ChunkedReaderImpl.cpp:3:0:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:131:9: error: 'headers_t' does not name a type
         headers_t headers;
         ^~~~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:132:9: error: 'args_t' does not name a type
         args_t args;
         ^~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:133:9: error: 'Proxy' does not name a type
         Proxy proxy;
         ^~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:136:5: error: 'virtual' outside class declaration
     virtual const Properties& GetProperties() const = 0;
     ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:136:47: error: non-member function 'const restc_cpp::Properties& restc_cpp::GetProperties( ' cannot have cv-qualifier
     virtual const Properties& GetProperties() const = 0;
                                               ^~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:136:55: error: function 'const restc_cpp::Properties& restc_cpp::GetProperties()' is initialized like a variable
     virtual const Properties& GetProperties() const = 0;
                                                       ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:137:5: error: 'virtual' outside class declaration
     virtual void SetProperties(Properties::ptr_t propreties) = 0;
     ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:137:64: error: function 'void restc_cpp::SetProperties(restc_cpp::Properties::ptr_t)' is initialized like a variable
     virtual void SetProperties(Properties::ptr_t propreties) = 0;
                                                                ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:140:5: error: 'virtual' outside class declaration
     virtual DataWriter& SendRequest(Context& ctx) = 0;
     ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:140:53: error: function 'restc_cpp::DataWriter& restc_cpp::SendRequest(restc_cpp::Context& ' is initialized like a variable
     virtual DataWriter& SendRequest(Context& ctx) = 0;
                                                     ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:152:5: error: 'virtual' outside class declaration
     virtual std::unique_ptr<Reply> GetReply(Context& ctx) = 0;
     ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:152:61: error: function 'std::unique_ptr<restc_cpp::Reply> restc_cpp::GetReply(restc_cpp::Context&)' is initialized like a variable
     virtual std::unique_ptr<Reply> GetReply(Context& ctx) = 0;
                                                             ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:160:5: error: 'virtual' outside class declaration
     virtual std::unique_ptr<Reply> Execute(Context& ctx) = 0;
     ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:160:60: error: function 'std::unique_ptr<restc_cpp::Reply> restc_cpp::Execute(restc_cpp::Context&)' is initialized like a variable
     virtual std::unique_ptr<Reply> Execute(Context& ctx) = 0;
                                                            ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:162:22: error: declaration of '~restc_cpp::Request' as non-member
     virtual ~Request() = default;
                      ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:166:18: error: 'Type' does not name a type
            const Type requestType,
                  ^~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:169:34: error: 'args_t' was not declared in this scope
            const boost::optional<args_t>& args = {},
                                  ^~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:169:40: error: template argument 1 is invalid
            const boost::optional<args_t>& args = {},
                                        ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:170:34: error: 'headers_t' was not declared in this scope
            const boost::optional<headers_t>& headers = {},
                                  ^~~~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:170:43: error: template argument 1 is invalid
            const boost::optional<headers_t>& headers = {},
                                           ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:171:34: error: 'auth_t' was not declared in this scope
            const boost::optional<auth_t>& auth = {});
                                  ^~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:171:40: error: template argument 1 is invalid
            const boost::optional<auth_t>& auth = {});
                                        ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:241:13: error: 'RestClient' does not name a type
     virtual RestClient& GetClient() = 0;
             ^~~~~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:261:44: error: 'Request' has not been declared
     virtual std::unique_ptr<Reply> Request(Request& req) = 0;
                                            ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:278:16: error: 'RestClient' has not been declared
                RestClient& rc);
                ^~~~~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:300:19: error: 'Request' does not name a type
     virtual const Request::Properties::ptr_t
                   ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:346:29: error: 'ConnectionPool' was not declared in this scope
     virtual std::shared_ptr<ConnectionPool> GetConnectionPool() = 0;
                             ^~~~~~~~~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:346:29: note: suggested alternative:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:57:7: note:   'restc_cpp::ConnectionPool'
 class ConnectionPool;
       ^~~~~~~~~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:346:43: error: template argument 1 is invalid
     virtual std::shared_ptr<ConnectionPool> GetConnectionPool() = 0;
                                           ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:360:32: error: 'Request' was not declared in this scope
         Create(boost::optional<Request::Properties> properties);
                                ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:360:32: note: suggested alternative:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:66:7: note:   'restc_cpp::Request'
 class Request {
       ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:360:51: error: template argument 1 is invalid
         Create(boost::optional<Request::Properties> properties);
                                                   ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:360:32: error: 'Request' was not declared in this scope
         Create(boost::optional<Request::Properties> properties);
                                ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:360:32: note: suggested alternative:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:66:7: note:   'restc_cpp::Request'
 class Request {
       ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:360:51: error: template argument 1 is invalid
         Create(boost::optional<Request::Properties> properties);
                                                   ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:360:32: error: 'Request' was not declared in this scope
         Create(boost::optional<Request::Properties> properties);
                                ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:360:32: note: suggested alternative:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:66:7: note:   'restc_cpp::Request'
 class Request {
       ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:360:51: error: template argument 1 is invalid
         Create(boost::optional<Request::Properties> properties);
                                                   ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:360:23: error: 'boost::optional' is not a type
         Create(boost::optional<Request::Properties> properties);
                       ^~~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:360:31: error: expected ',' or '...' before '<' token
         Create(boost::optional<Request::Properties> properties);
                               ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:365:44: error: 'Request' was not declared in this scope
         CreateUseOwnThread(boost::optional<Request::Properties> properties);
                                            ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:365:44: note: suggested alternative:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:66:7: note:   'restc_cpp::Request'
 class Request {
       ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:365:63: error: template argument 1 is invalid
         CreateUseOwnThread(boost::optional<Request::Properties> properties);
                                                               ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:365:44: error: 'Request' was not declared in this scope
         CreateUseOwnThread(boost::optional<Request::Properties> properties);
                                            ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:365:44: note: suggested alternative:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:66:7: note:   'restc_cpp::Request'
 class Request {
       ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:365:63: error: template argument 1 is invalid
         CreateUseOwnThread(boost::optional<Request::Properties> properties);
                                                               ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:365:44: error: 'Request' was not declared in this scope
         CreateUseOwnThread(boost::optional<Request::Properties> properties);
                                            ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:365:44: note: suggested alternative:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:66:7: note:   'restc_cpp::Request'
 class Request {
       ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:365:63: error: template argument 1 is invalid
         CreateUseOwnThread(boost::optional<Request::Properties> properties);
                                                               ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:365:35: error: 'boost::optional' is not a type
         CreateUseOwnThread(boost::optional<Request::Properties> properties);
                                   ^~~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:365:43: error: expected ',' or '...' before '<' token
         CreateUseOwnThread(boost::optional<Request::Properties> properties);
                                           ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:368:32: error: 'Request' was not declared in this scope
         Create(boost::optional<Request::Properties> properties,
                                ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:368:32: note: suggested alternative:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:66:7: note:   'restc_cpp::Request'
 class Request {
       ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:368:51: error: template argument 1 is invalid
         Create(boost::optional<Request::Properties> properties,
                                                   ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:368:32: error: 'Request' was not declared in this scope
         Create(boost::optional<Request::Properties> properties,
                                ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:368:32: note: suggested alternative:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:66:7: note:   'restc_cpp::Request'
 class Request {
       ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:368:51: error: template argument 1 is invalid
         Create(boost::optional<Request::Properties> properties,
                                                   ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:368:32: error: 'Request' was not declared in this scope
         Create(boost::optional<Request::Properties> properties,
                                ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:368:32: note: suggested alternative:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:66:7: note:   'restc_cpp::Request'
 class Request {
       ^~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:368:51: error: template argument 1 is invalid
         Create(boost::optional<Request::Properties> properties,
                                                   ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:368:23: error: 'boost::optional' is not a type
         Create(boost::optional<Request::Properties> properties,
                       ^~~~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:368:31: error: expected ',' or '...' before '<' token
         Create(boost::optional<Request::Properties> properties,
                               ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:368:9: error: 'static std::unique_ptr<RestClient> RestClient::Create(int)' cannot be overloaded
         Create(boost::optional<Request::Properties> properties,
         ^~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:360:9: error: with 'static std::unique_ptr<RestClient> RestClient::Create(int)'
         Create(boost::optional<Request::Properties> properties);
         ^~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h: In lambda function:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:332:52: error: no matching function for call to 'Context::Create(boost::asio::yield_context&, RestClient&)'
             auto ctx = Context::Create(yield, *this);
                                                    ^
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:277:9: note: candidate: static std::unique_ptr<Context> Context::Create(boost::asio::yield_context&, int&)
         Create(boost::asio::yield_context& yield,
         ^~~~~~
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:277:9: note:   no known conversion for argument 2 from 'RestClient' to 'int&'
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h: At global scope:
C:/msys64/home/gjacquemin/restc-cpp/include/restc-cpp/restc-cpp.h:379:1: error: expected declaration before '}' token
 } // restc_cpp
 ^
mingw32-make[2]: *** [src/CMakeFiles/restc-cpp.dir/build.make:64: src/CMakeFiles/restc-cpp.dir/ChunkedReaderImpl.cpp.obj] Error 1
mingw32-make[1]: *** [CMakeFiles/Makefile2:86: src/CMakeFiles/restc-cpp.dir/all] Error 2
mingw32-make: *** [Makefile:95: all] Error 2
```

--- a/include/restc-cpp/restc-cpp.h
+++ b/include/restc-cpp/restc-cpp.h
@@ -23,7 +23,7 @@
 #include "restc-cpp/helper.h"
 #include "restc-cpp/Connection.h"
 
-#ifdef _MSC_VER
+#if defined(_MSC_VER) || defined(__MINGW32__)
 // Thank you Microsoft for making every developers day productive
 #ifdef min
 #   undef min

